revising and condensing definitions (#5067)
fishtown-analytics/docs.getdbt.com: website/docs/reference/node-selection/test-selection-examples.md

</VersionBlock>

<IndirSelect features={'/snippets/indirect-selection-definitions.md'}/>

<!--tabs for eager mode, cautious mode, empty, and buildable mode -->
<!--Tabs for 1.5+ -->

### Indirect selection examples

<VersionBlock firstVersion="1.5">

> To visualize these methods, suppose you have `model_a`, `model_b`, and `model_c` and associated data tests. Here's what it will look like when you execute `dbt run` vs. `dbt build` with the various indirect selection modes:

Reviewer: ```suggestion
To visualize these methods, suppose you have `model_a`, `model_b`, and `model_c` and associated data tests. The following illustrates which tests will be run when you execute `dbt run` versus `dbt build` with the various indirect selection modes:
```

> To visualize these methods, suppose you have `model_a`, `model_b`, and `model_c` and associated data tests. Here's what it will look like when you execute `dbt run` vs. `dbt build` with the various indirect selection modes:

Reviewer 2: `dbt run` never selects or executes any tests, so it doesn't make sense to compare `dbt run` to `dbt build` in this context.

Rather, it only makes sense to vary:
1. Which models are selected (e.g., all vs. one).
1. Which indirect selection mode is chosen.

```suggestion
To visualize these methods, suppose you have `model_a`, `model_b`, and `model_c` and associated data tests. Here's what it will look like when you execute `dbt build` with the various indirect selection modes:
```

Reviewer 2: Became outdated, so added a new suggestion here:
https://github.com/dbt-labs/docs.getdbt.com/pull/5067/files#r1526934291

> To visualize these methods, suppose you have `model_a`, `model_b`, and `model_c` and associated data tests. The following illustrates which tests will be run when you execute `dbt run` versus `dbt build` with the various indirect selection modes:

Reviewer 2: `dbt run` never selects or executes any tests, so it doesn't make sense to compare `dbt run` to `dbt build` in this context.

Rather, it only makes sense to vary:
1. Which models are selected (e.g., all vs. one).
1. Which indirect selection mode is chosen.

So we should remove mention of `dbt run` entirely:

```suggestion
To visualize these methods, suppose you have `model_a`, `model_b`, and `model_c` and associated data tests. The following illustrates which tests will be run when you execute `dbt build` with the various indirect selection modes:
```

<DocCarousel slidesPerView={1}>

<Lightbox src src="/img/docs/reference/indirect-selection-dbt-run.png" width="85%" title="dbt run"/>

<Lightbox src src="/img/docs/reference/indirect-selection-eager.png" width="85%" title="Eager (default)"/>

<Lightbox src src="/img/docs/reference/indirect-selection-cautious.png" width="85%" title="Cautious"/>

<Lightbox src src="/img/docs/reference/indirect-selection-buildable.png" width="85%" title="Buildable"/>

<Lightbox src src="/img/docs/reference/indirect-selection-empty.png" width="85%" title="Empty"/>

</DocCarousel>

<Tabs queryString="indirect-selection-mode">
<TabItem value="eager" label="Eager mode (default)">

> For example, during the build process, any model that depends on the selected "orders" model or its dependent models will be built. Tests that refer to the selected "order" model or its dependent models will be executed during the testing process:

Reviewer 2: This isn't accurate.

Here's a slight tweak that gets it back in line:
```suggestion
In this example, during the build process, any test that depends on the selected "orders" model or its dependent models will be executed, even if it depends other models as well.
```

Author: That's much clearer too! 🙇🏻‍♀️

#### Cautious mode

  Ensures that tests are executed and models are built only when all necessary dependencies of the selected models are met. Restricts tests to only those that exclusively reference selected nodes. Tests will only be executed if all the nodes they depend on are selected, which prevents tests from running if one or more of its parents nodes are unselected and, consequently, unbuilt.

#### Buildable mode

  Only runs tests that refer to selected nodes (or their ancestors). Only builds models that rely on the selected nodes (or their ancestors). This mode is slightly more inclusive than "cautious" by including tests whose references are each within the selected nodes (or their ancestors). This mode is useful when a test depends on a model **and** a direct ancestor of that model, like confirming an aggregation has the same totals as its input.

#### Empty mode

  Restrict to tests that are only for the selected node and ignore all tests from the attached nodes 

> To visualize these methods, suppose you have `model_a`, `model_b`, and `model_c` and associated data tests. Here's what it will look like when you execute `dbt run` and then the various indirect selction modes:

Reviewer 3: ```suggestion
To visualize these methods, suppose you have `model_a`, `model_b`, and `model_c` and associated data tests. Here's what it will look like when you execute `dbt run` vs. `dbt build` with the various indirect selection modes:
```

<TabItem value="cautious" label="Cautious mode">

> For example, only the "orders" model (or node) and its dependent models will be selected for testing or building:

Reviewer 2: This isn't accurate.

This would get it back on track:
```suggestion
In this example, only tests that depend _exclusively_ on the "orders" model will be executed:
```

```shell
dbt test --select "orders" --indirect-selection=cautious
dbt build --select "orders" --indirect-selection=cautious

```

</TabItem>

<TabItem value="buildable" label="Buildable mode">

> For example, dbt runs tests that reference "orders" within the selected nodes (or their ancestors).

Reviewer 2: This looks accurate 👍 

But I'd still suggest clarifying that the above statement doesn't apply to dbt in general, and _only_ applies to the example below:

```suggestion
In this example, dbt executes tests that reference "orders" within the selected nodes (or their ancestors).
```

```shell
dbt test --select "orders" --indirect-selection=buildable
dbt build --select "orders" --indirect-selection=buildable
```

</TabItem>

<TabItem value="empty" label="Empty mode">

> This mode not execute any tests, whether they are directly attached to the selected node or not

Reviewer 2: This looks accurate 👍 

But looks like it is missing one word and the period:
```suggestion
This mode does not execute any tests, whether they are directly attached to the selected node or not.
```

Put another way, it will prevent tests from running if one or more of its parents is unselected.

```shell

dbt test --select "orders" --indirect-selection=cautious
dbt build --select "orders" --indirect-selection=cautious

```

</TabItem>

</Tabs>

</VersionBlock>

<!--End of tabs for eager mode, cautious mode, buildable mode, and empty mode -->

> ### Test selection syntax examples

Reviewer 2: Heads-up: changing the header will break any direct links to this section of the page.

So any GitHub issues that reference this section, internal links with our docs, or external links out on the internet will only go to the high-level page rather than redirecting to the new header :(

Author: Hrm good point

Author: I checked this, and in the last year, we have no direct hits to that sub section!

Author: Also noting that redirects don't work to subsections so I am just YOLOing and hoping I don't ruin any bookmarks.
